[3단계] 로또(2등)

RoadtoS7 opened the pull request:
꼼꼼한 피드백 감사합니다!! 이번 단계도 잘 부탁드립니다!!~ 🤗

코딩 컨벤션에 벗어나는 부분이 있거나, 혹은 반복해서 나타는 잘못된 코딩 습관이 있는지 피드백 부탁드립니다🙏

RoadtoS7 committed:
feat: InputView에 보너스 번호 입력 받는 메서드 추가

RoadtoS7 committed:
feat: 기존의 Lotto를 DefaultLotto로 만듬, Lotto 프로토콜 생성 및 WinningLotto가 Lotto 프로토콜을 따르도록 함

RoadtoS7 committed:
refactor: WinningCount 대신 Rank를 사용하도록 수정

RoadtoS7 committed:
feat: 당첨번호의 보너스번호를 갖는지 체크하는 WinningLotto.doesBonusNumberMathi()의 실패 테스트 케이스 추가

RoadtoS7 committed:
feat: WinningLotto.doesBonusNumberMath() 동작 구현

RoadtoS7 committed:
feat: LottoParser.parse -> LottoParser.parseLotto 로 이름 변경

RoadtoS7 committed:
feat: 사용자로부터 받은 입력으로 WinningLotto를 반환하는 parseWiningLotto 유닛 테스트 작성

RoadtoS7 committed:
feat: LottoParser.parseWinningLotto 구현

RoadtoS7 committed:
refactor: LottoParser.parseWinningLotto를 리팩토링

RoadtoS7 committed:
refactor: main에서 보너스 번호를 입력받아서 WinningLotto를 생성하도록 수정

RoadtoS7 committed:
feat: main에서 당첨번호, 보너스 번호를 입력받고, 수익률을 출력하도록 구현

RoadtoS7 committed:
feat: 보너스 번호가 일치해도 3등으로 계산되는 이슈 해결

- 이슈 원인
Rank의 from()에서 몇등인지 계산할 때 일치하는 번호의 개수가 5개이면 보너스 번호를 검사하지 않고 무조건 3번으로 분기 처리가 된다

- 해결 방법
Rank의 from()에서 switch 문의 case 순서를 변경하여 일치하는 번호의 개수가 5개일 때 보너스 번호까지 우선적으로 검사하도록 수정함

RoadtoS7 committed:
refactor: WinningLotto의 doesBonusNumberMatch() 메서드 이름을 checkBonusNumber(in:)으로 수정

RoadtoS7 committed:
style: 오타 수정

RoadtoS7 committed:
doc: 구현한 내용 README에 추가

RoadtoS7 commented on Lotto/Lotto/Lotto.swift:
```diff
@@ -14,47 +14,83 @@ enum LottoConstant {
+    func equalNumberCount(with lotto: Lotto) -> Int
+}
+
+extension Lotto {
+    func equalNumberCount(with lotto: Lotto) -> Int {
+        let count = lotto.numbers.filter {
+            self.numbers.contains($0)
+        }.count
         
-        var errorDescription: String? {
-            switch self {
-            case .invalidNumberCount(let count):
-                return "로또 번호 개수는 \(LottoConstant.numberCount)개입니다. 현재 입력된 로또 번호 개수: \(count)"
-                
+        return count
+    }
+}
+
```
WiningLotto(당첨번호와 보너스 번호를 갖는 로또), DefaultLotto(사용자가 구매한 로또)의 공통 구현사항을 Lotto 프로토콜을 이용하여 분리해보았습니다!

wonhee009 replied:
사용자가 구매한 로또니까 `DefaultLotto`보다는 `UserLotto`같이 사용자의 로또라는걸 알 수 있는 네이밍은 어때요?

RoadtoS7 replied:
[5a69780](https://github.com/next-step/swift-lotto/pull/25/commits/5a69780ba84410bb986851912236681e1eec4d30) 커밋에서 반영했습니다. 피드백 감사합니다!

RoadtoS7 commented on Lotto/Lotto/WinningRecord.swift:
```diff
@@ -9,37 +9,51 @@ import Foundation
     
     private let hundred: Double = 100
-    var value: [WinningCount: LottoCount] = [.three:0, .four: 0, .five: 0, .six: 0]
+    var value: [Rank: LottoCount] = [:]
     
-    init(equalNumberCounts: Array<Int>) {
-        equalNumberCounts.compactMap { WinningCount(rawValue: $0) }
-                        .forEach { value[$0]! += 1 }
+    init(recordUnits: [RecordUnit]) {
+        Rank.allCases.forEach { self.value[$0] = 0 }
+        
+        recordUnits.map{ matchingCount, matchBonus in
+            Rank.from(matchingCount: matchingCount, matchBonus: matchBonus)
+        }.forEach { value[$0]! += 1 }
     }
     
     func calculateProfitRate(inputMoney: Money) -> Double {
         let profit = value.reduce(0) { (partialResult: Int, value) in
-            let (winningCount, lottoCount) = value
-            return partialResult + winningCount.prize * lottoCount
+            let (rank, lottoCount) = value
+            return partialResult + rank.prize * lottoCount
         }
         
         return Double(profit) / Double(inputMoney.value) * hundred
```
WinningRecord의 initializer에서 필요한 파라미터를 typealias로 구현하는 것을 처음으로 시도해보았습니다!
typealias를 사용했을 때 발생할 수 있는 문제점이 있는지 궁금합니다!

wonhee009 replied:
typealias 사용해보셨군요 👍 
이후에 스펙이 수정되어 RecordUnit에 포함되어야 하는 요소가 많아질때가 단점일거 같습니다.
```swift
typealias Semantic = SemanticColor
```
자동완성을 사용한다면 `Semantic`, `SemanticColor` 모두 걸려서 코드 통일성에서 단점이 될 수도 있을거 같구요.

wonhee009 commented on Lotto/Lotto/Lotto.swift:
```diff
@@ -14,47 +14,83 @@ enum LottoConstant {
-                return "로또 번호 개수는 \(LottoConstant.numberCount)개입니다. 현재 입력된 로또 번호 개수: \(count)"
-                
+        return count
+    }
```
`Lotto`의 extension으로 있으면 WinningLotto에서도 equal을 호출할 수 있고 DefaultLotto에서도 equal을 호출 할 수 있을텐데요.
두가지 타입에서 모두 호출할 수 있다면 혼란을 줄 수 있지 않을까요?
DefaultLotto나 WinningLotto의 Extension으로 제한하는게 좋지 않을까 합니다.

RoadtoS7 replied:
번호 비교는 보통 WinningLotto(당첨번호)하고 이루어지기 때문에, WinningLotto의 extension으로 제한하는 것으로 수정했습니다!
[00202e6](https://github.com/next-step/swift-lotto/pull/25/commits/00202e60f5dcd6a771f4d6db373902c1684e02da) 커밋에서 반영했습니다. 피드백 감사합니다!

wonhee009 commented on Lotto/Lotto/Lotto.swift:
```diff
@@ -14,47 +14,83 @@ enum LottoConstant {
+            return "로또 번호는 \(LottoConstant.minValue)~\(LottoConstant.maxValue) 이내여야 합니다."
         }
     }
+}
```
`Lotto`라는 하나의 파일로 분리하고, `WinningLotto`, `DefaultLotto`도 다른 파일로 분리할 필요 없을까요?

RoadtoS7 replied:
[102ebfa](https://github.com/next-step/swift-lotto/pull/25/commits/102ebfa2d0c6e1111bd2bb49c19e8284f36c812e) 커밋에서 반영했습니다! 피드백 감사합니다!

wonhee009 commented on Lotto/Lotto/Lotto.swift:
```diff
@@ -14,47 +14,83 @@ enum LottoConstant {
+    func checkBonusNumber(in lotto: Lotto) -> Bool {
+        let count = lotto.numbers.filter { $0 == bonusNumber }.count
+        return count > 0
+    }
```
보너스 숫자 검증을 외부에서 호출하는가요?

RoadtoS7 replied:
피드백 감사합니다! 
https://github.com/next-step/swift-lotto/pull/25#discussion_r874937318  이 리뷰에 피드백 반영한 내용을 표시해두고, 궁금한 점을 올려두었습니다. 확인부탁드립니다 🙏

wonhee009 commented on Lotto/Lotto/Lotto.swift:
```diff
@@ -14,47 +14,83 @@ enum LottoConstant {
+        }
+        if hasValidNumbers == false {
+            throw LottoError.invalidLottoNumber
+        }
```
로또 숫자 검증하는 부분이 Winning, Default 동일하게 있는거 같은데요.
Validate struct로 하나 빼는 것도 좋을거 같습니다 :)

RoadtoS7 replied:
피드백 감사합니다! 피드백을 아래 2가지 방법으로 구현해본 후, 2번 방법(extension)을 사용한 방법이 더 코드가 깔끔한 것 같아서 2번을 최종적으로 선택했습니다!
원희님의 생각은 어떠신지 궁금합니다!

1. 로또 숫자를 검증하는 Util 성 구조체인 `LottoNumberValidator` 를 사용
[LottoNumberValidator 파일](https://github.com/next-step/swift-lotto/blob/9ef4ad5034f2251fe02b01ed39fe9db5c6f1d00a/Lotto/Lotto/Lotto/LottoNumberValidator.swift)
[LottoNumberValidator를 사용한 WinningLotto](https://github.com/next-step/swift-lotto/blob/6dbe82f273d126d8b443de86cdb1419b7f2657d7/Lotto/Lotto/Lotto/WinningLotto.swift)

2. Collection Extension으로 `canBeLottoNumbers()` 메서드를 구현하여 사용
[Colleciton Extension](https://github.com/next-step/swift-lotto/blob/9ef4ad5034f2251fe02b01ed39fe9db5c6f1d00a/Lotto/Lotto/Extension/CollectionExtension.swift)
[`canBeLottoNumbers` 를 사용한 WinningLotto](https://github.com/next-step/swift-lotto/blob/9ef4ad5034f2251fe02b01ed39fe9db5c6f1d00a/Lotto/Lotto/Lotto/WinningLotto.swift)

wonhee009 replied:
저는 1번 방법이 좀 더 맞는거 같습니다.
Collection extension에 validate가 구현되어 있으면 이후 Lotto는 아니지만 Int Collection인 타입에서도 해당 검증 메서드를 호출할 수 있을거 같습니다.

RoadtoS7 replied:
피드백 감사합니다!
[bbaa551](https://github.com/next-step/swift-lotto/pull/25/commits/bbaa5517fb2683c90b25c332779a75be0bf926af)  커밋에서 반영했습니다!

wonhee009 commented on Lotto/Lotto/LottoFactory.swift:
```diff
@@ -10,14 +10,14 @@ import Foundation
 struct LottoFactory {
```
`LottoFactory`인데 DefaultLotto만 생성하나용?.? 이름상 WinningLotto도 생성할거 같아서요!

RoadtoS7 replied:
[f544752](https://github.com/next-step/swift-lotto/pull/25/commits/f544752b9330b848f34f5806035e4836125400cc) 에서 UserLottoFactory로 이름 수정하였습니다! 피드백 감사합니다!

wonhee009 commented on Lotto/Lotto/Util/WinningRecordFormatter.swift:
```diff
@@ -9,8 +9,12 @@ import Foundation
-        return WinningCount.allCases.reduce("") { partialResult, winningCount in
-            partialResult + "\(winningCount.rawValue)개 일치(\(winningCount.prize))원 - \(statistics.value[winningCount]!)개 \n"
-        }
+        return "3개 일치 (5000원)- \(statistics.value[.fifth]!)개 \n"
+        + "4개 일치 (50000원)- \(statistics.value[.fourth]!)개 \n"
+        + "5개 일치 (50000원)- \(statistics.value[.third]!)개 \n"
+        + "5개 일치, 보너스 번호 일치(30000000원) - \(statistics.value[.second]!)개 \n"
+        + "6개 일치 (50000원)- \(statistics.value[.first]!)개 \n"
```
```swift
struct WinningRecordFormatter {
   private enum Text {
      static let first = "3개 일치 (5000원)- %@개"
      ...
   }

   static func format(_ statistics: WinningRecord) -> String {
      let firstString = String(format: Text.first, statistics.value[.fifth])
      ...
   }
}
```

이런식으로 사용하는건 어떤가요?

RoadtoS7 replied:
오 제안해주신 방법이 훨씬 깔끔한 것 같습니다! 
피드백 감사합니다!! [95cc098](https://github.com/next-step/swift-lotto/pull/25/commits/95cc09800236d9921aef08950b83cfb4b562c48c) 커밋에서 반영했습니다.

wonhee009 reviewed (changes requested):
수고하셨습니다 🎉 
코멘트 남겨두었으니 확인 부탁드려요 🙏

RoadtoS7 committed:
refactor: DefaultLotto -> UserLotto 네이밍 수정

RoadtoS7 committed:
refactor: Lotto, DefaultLotto, WinningLotto를 별도의 파일로 분리

RoadtoS7 committed:
refactor: equalNumberCount(with:)을 WinningLotto에만 존재하도록 수정

equalNumberCount를 Lotto 프로토콜의 구성요소로 두고, default 구현을 제공했었다.
하지만 WinningLotto(당첨로또)와 일치하는 번호를 구하기 위해 필요한 기능이므로, WinningLotto에만 equalNumberCount(with:)가 존재하도록 수정했다.

RoadtoS7 committed:
refactor: LottoFactory -> UserLottoFactory로 이름 수정

RoadtoS7 committed:
refactor: WinningLotto, UserLotto, LottoFactory를 Lotto 디렉토리로 폴더링

RoadtoS7 committed:
feat: LottoNumber의 규칙을 만족하는지 체크하는 LottoNumberValidator 유닛 테스트 구현

RoadtoS7 committed:
feat: LottNumberValidator.validate(numbers:) 구현

RoadtoS7 committed:
feat: WinningLotto, UserLotto에서 LottoNumberValidator를 사용하도록 수정

RoadtoS7 committed:
refactor: LottoNumberValidator의 기능을 Collection Extension의 canBeLottoNumbers() 메서드로 구현 및 적용

RoadtoS7 committed:
refactor: WinningRecordFormatter에서 출력 String을 enum Text로 wrapping함

- 수정 이유: 하드 코딩을 지양하자!

RoadtoS7 committed:
feat: UserLotto.contains(특정 번호를 가지면 true를 반환) 메서드 유닛 테스트를 UserLottoTest에 추가

RoadtoS7 committed:
feat: UserLotto.contains(특정 번호를 가지면 true를 반환) 메서드 구현

RoadtoS7 committed:
feat: UserLotto.contains(특정 번호를 가지면 true를 반환) 메서드 실패하는 유닛 테스트 추가

RoadtoS7 committed:
refactor: Lotto에 contains(number:)메서드(number를 가지면 true를 반환) 추가

RoadtoS7 committed:
refactor: WinningLotto의 equalNumberCounts(with:)에서 Lotto.contains(number:)사용하도록 수정

RoadtoS7 commented on Lotto/Lotto/Lotto/WinningLotto.swift:
```diff
@@ -0,0 +1,43 @@
+        self.bonusNumber = bonusNumber
+    }
+    
+    func checkBonusNumber(in lotto: Lotto) -> Bool {
+        return lotto.contains(number: bonusNumber)
+    }
```
lotto의 numbers 프로퍼티를 WinningLotto에서 참조하는 부분을 지양하는 것을 피드백해주신 걸로 이해했는데 제가 제대로 이해한 것이 맞는지 궁금합니다!
우선 lotto의 numbers를 직접 참조하지 않도록 contains(number:) 메서드를 추가로 구현했습니다!

wonhee009 replied:
앗ㅋㅋ저는 private일 필요가 없는지 궁금했던거였습니다!ㅋㅋㅋ

RoadtoS7 replied:
앗 그런 의미였군요..!
LottoBag에서 이 메서드를 호출하고 있어서 외부에서도 참조할 수 있도록 해두었습니다..!

RoadtoS7 committed:
refactor: main에서 변수 lottoTicket 이름 -> lottoBag으로 수정

wonhee009 commented on Lotto/Lotto/View/OutputView.swift:
```diff
@@ -23,7 +23,7 @@ struct OutputView {
     
     static func print(profitRate: Double) {
-        let profiteRateView = String(format: "총 수익률은 %2.f입니다.", profitRate)
+        let profiteRateView = String(format: "총 수익률은 %.2f입니다.", profitRate)
```
👍

wonhee009 commented on Lotto/Lotto/Util/WinningRecordFormatter.swift:
```diff
@@ -8,13 +8,20 @@
     
-    
+    static func format(_ statistics: WinningRecord) -> String {
+        return String(format: Text.fifth, statistics.value[.first]!) + newLine
+                + String(format: Text.fourth, statistics.value[.fourth]!) + newLine
+                + String(format: Text.third, statistics.value[.third]!) + newLine
+                + String(format: Text.second, statistics.value[.second]!) + newLine
+                + String(format: Text.first, statistics.value[.first]!)
```
```swift
let fifthText = String(format: Text.fifth, statistics.value[.first]!) + newLine
let fourthText = String(format: Text.fourth, statistics.value[.fourth]!) + newLine
...

return fifthText + fourthText ...
```
저는 이렇게 사용하는 편이긴 합니다 :)
이 부분은 스타일 차이여서 그냥 의견 정도로 봐주시면 좋을거 같습니다!

RoadtoS7 replied:
[22583f3](https://github.com/next-step/swift-lotto/pull/25/commits/22583f38b0d2ec0ff58bd104022f3be2ea6d41c2) 커밋에서 반영해보았습니다! 피드백 감사합니다!

RoadtoS7 committed:
refactor: UserLotto, WinningLotto의 intializer에서 로또 번호가 될 수 있는지 체크할 때 LottoNumberValidator를 사용하도록 수정

RoadtoS7 committed:
fix: calculateProfitRate(inputMoney:)의 결과로 백분율이 나오는 이슈 해결

RoadtoS7 committed:
refactor: LottoParseer가 ,를 기준으로도 파싱되도록 수정

RoadtoS7 committed:
fix: OutputView에서 LocalizedError의 errorDescription을 출력할 수 있도록 수정

RoadtoS7 committed:
refactor: WinningRecordFormatter에서 출력할 String을 변수에 할당에서 사용

wonhee009 reviewed (approved):
수고하셨습니다 😄
